### PR TITLE
provider/aws: API Gateway Custom Authorizer

### PR DESCRIPTION
**Pull request by radeksimko**
Replaces https://github.com/hashicorp/terraform/pull/6731

Thanks @johnjelinek for most of the work here. I just added a few missing bits.

### Test plan
One test may _actually_ fail until https://github.com/hashicorp/terraform/pull/8533 is merged.

```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=AWSAPIGateway'
```
```
TF_ACC=1 go test ./builtin/providers/aws -v -run=AWSAPIGateway -timeout 120m
=== RUN   TestAccAWSAPIGatewayAccount_importBasic
--- PASS: TestAccAWSAPIGatewayAccount_importBasic (16.81s)
=== RUN   TestAccAWSAPIGatewayApiKey_importBasic
--- PASS: TestAccAWSAPIGatewayApiKey_importBasic (43.76s)
=== RUN   TestAccAWSAPIGatewayAccount_basic
--- PASS: TestAccAWSAPIGatewayAccount_basic (120.88s)
=== RUN   TestAccAWSAPIGatewayApiKey_basic
--- PASS: TestAccAWSAPIGatewayApiKey_basic (42.53s)
=== RUN   TestAccAWSAPIGatewayAuthorizer_basic
--- PASS: TestAccAWSAPIGatewayAuthorizer_basic (52.22s)
=== RUN   TestAccAWSAPIGatewayBasePath_basic
--- PASS: TestAccAWSAPIGatewayBasePath_basic (40.37s)
=== RUN   TestAccAWSAPIGatewayDeployment_basic
--- PASS: TestAccAWSAPIGatewayDeployment_basic (37.23s)
=== RUN   TestAccAWSAPIGatewayDomainName_basic
--- PASS: TestAccAWSAPIGatewayDomainName_basic (20.89s)
=== RUN   TestAccAWSAPIGatewayIntegrationResponse_basic
--- PASS: TestAccAWSAPIGatewayIntegrationResponse_basic (58.66s)
=== RUN   TestAccAWSAPIGatewayIntegration_basic
--- PASS: TestAccAWSAPIGatewayIntegration_basic (48.09s)
=== RUN   TestAccAWSAPIGatewayMethodResponse_basic
--- PASS: TestAccAWSAPIGatewayMethodResponse_basic (55.87s)
=== RUN   TestAccAWSAPIGatewayMethod_basic
--- PASS: TestAccAWSAPIGatewayMethod_basic (43.93s)
=== RUN   TestAccAWSAPIGatewayMethod_customauthorizer
--- PASS: TestAccAWSAPIGatewayMethod_customauthorizer (59.41s)
=== RUN   TestAccAWSAPIGatewayModel_basic
--- PASS: TestAccAWSAPIGatewayModel_basic (24.93s)
=== RUN   TestAccAWSAPIGatewayResource_basic
--- PASS: TestAccAWSAPIGatewayResource_basic (29.13s)
=== RUN   TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (43.80s)
PASS
ok     	github.com/hashicorp/terraform/builtin/providers/aws   	738.540s
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_api_gateway_method.go [hunk 1]
```diff
@@ -107,6 +112,7 @@ func resourceAwsApiGatewayMethodCreate(d *schema.ResourceData, meta interface{})
 		RequestModels:     aws.StringMap(models),
 		RequestParameters: aws.BoolMap(parameters),
 		ApiKeyRequired:    aws.Bool(d.Get("api_key_required").(bool)),
+		AuthorizerId:      aws.String(d.Get("authorizer_id").(string)),
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_api_gateway_method.go]: If this is optional, shouldn't we check that it is present before passing it to the API?

Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_method.go]: True, we could do that, although based on tests it has the same effect as passing an empty string. Admittedly I'm in favour of changes that make resulting HTTP requests smaller so I'll update the PR. 😉 👍

**Comment by radeksimko** [comment]
PR Updated + rebased. Also I reran the acceptance tests and they're all passing now. 🎉

**Comment by stack72** [comment]
@radeksimko 

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=AWSAPIGateway'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=AWSAPIGateway -timeout 120m
=== RUN   TestAccAWSAPIGatewayAccount_importBasic
--- PASS: TestAccAWSAPIGatewayAccount_importBasic (33.40s)
=== RUN   TestAccAWSAPIGatewayApiKey_importBasic
--- PASS: TestAccAWSAPIGatewayApiKey_importBasic (39.00s)
=== RUN   TestAccAWSAPIGatewayAccount_basic
--- PASS: TestAccAWSAPIGatewayAccount_basic (99.72s)
=== RUN   TestAccAWSAPIGatewayApiKey_basic
--- PASS: TestAccAWSAPIGatewayApiKey_basic (39.76s)
=== RUN   TestAccAWSAPIGatewayAuthorizer_basic
--- PASS: TestAccAWSAPIGatewayAuthorizer_basic (51.75s)
=== RUN   TestAccAWSAPIGatewayBasePath_basic
--- PASS: TestAccAWSAPIGatewayBasePath_basic (37.72s)
=== RUN   TestAccAWSAPIGatewayDeployment_basic
--- PASS: TestAccAWSAPIGatewayDeployment_basic (33.24s)
=== RUN   TestAccAWSAPIGatewayDomainName_basic
--- PASS: TestAccAWSAPIGatewayDomainName_basic (20.19s)
=== RUN   TestAccAWSAPIGatewayIntegrationResponse_basic
--- PASS: TestAccAWSAPIGatewayIntegrationResponse_basic (51.22s)
=== RUN   TestAccAWSAPIGatewayIntegration_basic
--- PASS: TestAccAWSAPIGatewayIntegration_basic (43.42s)
=== RUN   TestAccAWSAPIGatewayMethodResponse_basic
--- PASS: TestAccAWSAPIGatewayMethodResponse_basic (44.99s)
=== RUN   TestAccAWSAPIGatewayMethod_basic
--- PASS: TestAccAWSAPIGatewayMethod_basic (38.66s)
=== RUN   TestAccAWSAPIGatewayMethod_customauthorizer
--- PASS: TestAccAWSAPIGatewayMethod_customauthorizer (56.44s)
=== RUN   TestAccAWSAPIGatewayModel_basic
--- PASS: TestAccAWSAPIGatewayModel_basic (20.66s)
=== RUN   TestAccAWSAPIGatewayResource_basic
--- PASS: TestAccAWSAPIGatewayResource_basic (24.99s)
=== RUN   TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (39.83s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	675.007s
```

File: website/source/docs/providers/aws/r/api_gateway_method.html.markdown [hunk 2]
```diff
@@ -39,7 +39,8 @@ The following arguments are supported:
 * `rest_api_id` - (Required) The ID of the associated REST API
 * `resource_id` - (Required) The API resource ID
 * `http_method` - (Required) The HTTP Method (`GET`, `POST`, `PUT`, `DELETE`, `HEAD`, `OPTION`)
-* `authorization` - (Required) The type of authorization used for the method
+* `authorization` - (Required) The type of authorization used for the method (`NONE`, `CUSTOM`)
```
Comment (swestcott) [inline website/source/docs/providers/aws/r/api_gateway_method.html.markdown]: AWS_IAM is also a valid authorisation type (authorizer_id not required for this)

Comment (radeksimko) [inline website/source/docs/providers/aws/r/api_gateway_method.html.markdown]: Thanks for the note @swestcott , would you mind sending a PR with this small fix?

**Pull request merged**